File: Week_06/kotlin/src/_63_UniquePathsII.kt
//这里是一个错误示范。。。
fun uniquePathsWithObstacles_fail(obstacleGrid: Array<IntArray>): Int {
    //f(i, j)走到i,j的路径数量
    //f(i, j) = f(i - 1, j) + f(i, j - 1)
    if (obstacleGrid.isEmpty() || obstacleGrid[0].isEmpty()) return 0
    if (obstacleGrid[0][0] == 1) return 0
    val m = obstacleGrid.size
    val n = obstacleGrid[0].size
    val dp = IntArray(n + 1)
    dp[0] = 1
    for (i in 1..m) {
        for (j in 1..n) {
            if (obstacleGrid[i - 1][j - 1] == 0)
                dp[j] = dp[j - 1] + dp[j]
            else
                dp[j] = 0
        }
    }
    return dp[n]
}

fun uniquePathsWithObstacles(obstacleGrid: Array<IntArray>): Int {
    //f(i, j)走到i,j的路径数量
    //f(i, j) = f(i - 1, j) + f(i, j - 1)
    if (obstacleGrid.isEmpty() || obstacleGrid[0].isEmpty()) return 0
    if (obstacleGrid[0][0] == 1) return 0
    val m = obstacleGrid.size
    val n = obstacleGrid[0].size
    val dp = IntArray(n)
    dp[0] = 1
    for (i in 0 until m) {
        for (j in 0 until n) {
            if (obstacleGrid[i][j] == 1) {
                dp[j] = 0
            } else if (j > 0) {
                dp[j] += dp[j - 1]
            }
        }
    }
    return dp[n - 1]
}

File: Week_06/cpp/src/_63_UniquePathsII.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int uniquePathsWithObstacles(vector<vector<int>> &obstacleGrid)
    {
        // f(i, j) = f(i - 1, j) + f(i, j - 1)
        const int m = obstacleGrid.size(), n = obstacleGrid[0].size();
        vector<int> dp(n);
        for (int j = 0; j < n; j++)
        {
            if (obstacleGrid[0][j])
                break;
            dp[j] = 1;
        }
        for (int i = 1; i < m; i++)
        {
            if (obstacleGrid[i][0])
                dp[0] = 0;
            for (int j = 1; j < n; j++)
            {
                dp[j] = obstacleGrid[i][j] ? 0 : dp[j] + dp[j - 1];
            }
        }
        return dp[n - 1];
    }
};